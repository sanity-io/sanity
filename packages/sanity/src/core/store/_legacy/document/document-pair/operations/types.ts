import {SanityDocument, Schema} from '@sanity/types'
import {SanityClient} from '@sanity/client'
import {IdPair} from '../../types'
import {DocumentVersionSnapshots} from '../snapshotPair'
import {HistoryStore} from '../../../history'

/** @internal */
export interface OperationImpl<
  ExtraArgs extends any[] = [],
  DisabledReason extends string = string,
> {
  disabled: (args: OperationArgs) => DisabledReason | 'NOT_READY' | false
  execute(args: OperationArgs, ...extra: ExtraArgs): void
}

/** @internal */
export interface Operation<ExtraArgs extends any[] = [], ErrorStrings extends string = string> {
  disabled: false | ErrorStrings | 'NOT_READY'
  execute(...extra: ExtraArgs): void
}

type GuardedOperation = Operation<any[], 'NOT_READY'>
type Patch = any

/** @internal */
// Note: Changing this interface in a backwards incompatible manner will be a breaking change
export interface OperationsAPI {
  commit: Operation | GuardedOperation
  delete: Operation<[], 'NOTHING_TO_DELETE' | 'NOT_READY'>
  del: Operation<[], 'NOTHING_TO_DELETE'> | GuardedOperation
  publish:
    | Operation<[], 'LIVE_EDIT_ENABLED' | 'ALREADY_PUBLISHED' | 'NO_CHANGES'>
    | GuardedOperation
  patch: Operation<[patches: Patch[], initialDocument?: Record<string, any>]> | GuardedOperation
  discardChanges: Operation<[], 'NO_CHANGES' | 'NOT_PUBLISHED'> | GuardedOperation
  unpublish: Operation<[], 'LIVE_EDIT_ENABLED' | 'NOT_PUBLISHED'> | GuardedOperation
  duplicate: Operation<[documentId: string], 'NOTHING_TO_DUPLICATE'> | GuardedOperation
  restore: Operation<[revision: string]> | GuardedOperation
}

/** @internal */
export interface OperationArgs {
  historyStore: HistoryStore
  client: SanityClient
  schema: Schema
  typeName: string
  idPair: IdPair
  snapshots: {draft: null | SanityDocument; published: null | SanityDocument}
  draft: DocumentVersionSnapshots
  published: DocumentVersionSnapshots
}
