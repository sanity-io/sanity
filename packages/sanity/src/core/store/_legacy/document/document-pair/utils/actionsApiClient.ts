import {type SanityClient} from 'sanity'

const ACTIONS_API_MINIMUM_VERSION = '2024-05-23'

export function actionsApiClient(client: SanityClient): SanityClient {
  return client.withConfig({
    apiVersion: ACTIONS_API_MINIMUM_VERSION,
  })
}
