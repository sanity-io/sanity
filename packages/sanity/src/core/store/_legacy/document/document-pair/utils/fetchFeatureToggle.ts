import {type SanityClient} from '@sanity/client'
import {map, type Observable, of, ReplaySubject, timeout, timer} from 'rxjs'
import {catchError, concatMap, share} from 'rxjs/operators'

interface ActionsFeatureToggle {
  actions: boolean
}
const CACHE = new WeakMap<SanityClient, Observable<boolean>>()

// How often to refresh the feature toggle
const REFRESH_INTERVAL = 1000 * 120

// Timer used to reset the observable when it completes or it's refcount drops to zero
const RESET_TIMER = timer(REFRESH_INTERVAL)

function createFeatureToggle(client: SanityClient) {
  const {dataset} = client.config()

  return timer(0, REFRESH_INTERVAL).pipe(
    concatMap(() =>
      client.observable
        .request({
          uri: `/data/actions/${dataset}`,
          withCredentials: true,
        })
        .pipe(
          map((res: ActionsFeatureToggle) => res.actions),
          timeout({first: 2000, with: () => of(false)}),
          catchError(() =>
            // If we fail to fetch the feature toggle, we'll just assume it's disabled and fallback to legacy mutations
            of(false),
          ),
        ),
    ),
    share({
      // replay latest known state to new subscribers
      connector: () => new ReplaySubject(1),
      // this will typically be completed and unsubscribed from right after the answer is received, so we don't want to reset
      resetOnComplete: () => RESET_TIMER,
      // keep it alive for some time after the last subscriber unsubscribes
      resetOnRefCountZero: () => RESET_TIMER,
    }),
  )
}

export const fetchFeatureToggle = (client: SanityClient): Observable<boolean> => {
  if (!CACHE.has(client)) {
    CACHE.set(client, createFeatureToggle(client))
  }
  return CACHE.get(client)!
}
