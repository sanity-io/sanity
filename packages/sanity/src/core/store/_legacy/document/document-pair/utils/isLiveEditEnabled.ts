import {type Schema} from '@sanity/types'

export const isLiveEditEnabled = (schema: Schema, typeName: string): boolean =>
  (globalThis as any).__SECRET_INTERNALS_DO_NOT_USE_OR_YOU_WILL_BE_FIRED_LIVE_EDIT_OVERRIDE ??
  schema.get(typeName)?.liveEdit === true
