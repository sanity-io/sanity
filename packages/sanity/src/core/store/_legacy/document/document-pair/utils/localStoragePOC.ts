import {isSanityDocument, type SanityDocument} from '@sanity/types'

import {supportsLocalStorage} from '../../../../../util/supportsLocalStorage'
import {type IdPair} from '../../types'

const createDocumentLocalStorageKey = (documentId: string) => `sanity:editState:${documentId}`

const getDocumentFromLocalStorage = (id: string): SanityDocument | null => {
  if (!supportsLocalStorage) return null

  const key = createDocumentLocalStorageKey(id)

  try {
    const document = localStorage.getItem(key)

    if (!document) return null
    const parsed = JSON.parse(document)
    return isSanityDocument(parsed) ? parsed : null
  } catch (error) {
    console.error(`Error parsing document with ID ${id} from localStorage:`, error)
    return null
  }
}

const saveDocumentToLocalStorage = (document: SanityDocument) => {
  if (!supportsLocalStorage) return

  const key = createDocumentLocalStorageKey(document._id)

  try {
    localStorage.setItem(key, JSON.stringify(document))
  } catch (error) {
    console.error(`Error saving document with ID ${document._id} to localStorage:`, error)
  }
}

/**
 * Function to get the draft and published document from local storage
 * it's not production ready, it's POC only, local storage supports up to 5mb of data which won't be enough for the datasets.
 * @internal
 * @hidden
 */
export const getPairFromLocalStorage = (idPair: IdPair) => {
  if (!supportsLocalStorage) {
    return {
      draft: null,
      published: null,
    }
  }

  return {
    draft: getDocumentFromLocalStorage(idPair.draftId),
    published: getDocumentFromLocalStorage(idPair.publishedId),
  }
}

/**
 * Function to save the draft and published documents to local storage.
 * Note: This is a proof of concept and not production-ready.
 * Local storage supports up to 5MB of data, which will not be sufficient for large datasets.
 * @internal
 * @hidden
 */
export const savePairToLocalStorage = (
  documentPair: {
    draft: SanityDocument | null
    published: SanityDocument | null
  } | null,
) => {
  if (documentPair?.draft) {
    saveDocumentToLocalStorage(documentPair.draft)
  }
  if (documentPair?.published) {
    saveDocumentToLocalStorage(documentPair.published)
  }
}
