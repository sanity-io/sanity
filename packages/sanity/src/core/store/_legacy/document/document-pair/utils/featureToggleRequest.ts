//this is an example function -- we're just calling a document

import {type SanityClient} from '@sanity/client'
import {map, type Observable, of} from 'rxjs'

import {type IdPair} from '../../types'
import {memoize} from '../../utils/createMemoizer'

//in the "real" code, this would be observable.request, to a URI
export const featureToggleRequest: (
  client: SanityClient,
  idPair: IdPair,
  serverActionsEnabled: boolean,
) => Observable<boolean> = memoize(
  (client: SanityClient, idPair: IdPair, serverActionsEnabled: boolean): Observable<boolean> => {
    if (!serverActionsEnabled) {
      return of(false)
    }
    return client.observable
      .fetch('*[_id == $id][0]', {id: '20449512-1e9c-44f0-a509-417c901fbbbd'})
      .pipe(map((doc) => doc.name.split('CONTROL FOR SERVER ACTIONS: ')[1] === 'enabled'))
  },
  (client: SanityClient, idPair: IdPair, serverActionsEnabled: boolean) => {
    const config = client.config()
    return `${config.dataset ?? ''}-${config.projectId ?? ''}-${idPair.publishedId}-${serverActionsEnabled ? '-serverActionsEnabled' : ''}`
  },
)
