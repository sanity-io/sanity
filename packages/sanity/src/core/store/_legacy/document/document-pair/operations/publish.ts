import {omit} from 'lodash'
import {isReference} from '@sanity/types'
import {isLiveEditEnabled} from '../utils/isLiveEditEnabled'
import {OperationImpl} from './index'

function strengthenOnPublish<T = any>(obj: T): T {
  if (isReference(obj)) {
    if (obj._strengthenOnPublish) {
      return omit(
        obj,
        ['_strengthenOnPublish'].concat(obj._strengthenOnPublish.weak ? [] : ['_weak']),
      ) as T
    }
    return obj
  }
  if (typeof obj !== 'object' || !obj) return obj
  if (Array.isArray(obj)) return obj.map(strengthenOnPublish) as T
  return Object.fromEntries(
    Object.entries(obj).map(([key, value]) => [key, strengthenOnPublish(value)] as const),
  ) as T
}

type DisabledReason = 'LIVE_EDIT_ENABLED' | 'ALREADY_PUBLISHED' | 'NO_CHANGES'

export const publish: OperationImpl<[], DisabledReason> = {
  disabled: ({schema, typeName, snapshots}) => {
    if (isLiveEditEnabled(schema, typeName)) {
      return 'LIVE_EDIT_ENABLED'
    }
    if (!snapshots.draft) {
      return snapshots.published ? 'ALREADY_PUBLISHED' : 'NO_CHANGES'
    }
    return false
  },
  execute: ({client, idPair, snapshots}) => {
    if (!snapshots.draft) {
      throw new Error('cannot execute "publish" when draft is missing')
    }
    const value = strengthenOnPublish(omit(snapshots.draft, '_updatedAt'))
    const tx = client.observable.transaction()
    if (!snapshots.draft) {
      throw new Error('cannot execute "publish" when draft is missing')
    }

    if (snapshots.published) {
      // If it exists already, we only want to update it if the revision on the remote server
      // matches what our local state thinks it's at
      tx.patch(idPair.publishedId, {
        // Hack until other mutations support revision locking
        unset: ['_revision_lock_pseudo_field_'],
        ifRevisionID: snapshots.published._rev,
      })

      tx.createOrReplace({
        ...value,
        _id: idPair.publishedId,
        _type: snapshots.draft._type,
      })
    } else {
      // If the document has not been published, we want to create it - if it suddenly exists
      // before being created, we don't want to overwrite if, instead we want to yield an error
      tx.create({
        ...value,
        _id: idPair.publishedId,
        _type: snapshots.draft._type,
      })
    }

    tx.delete(idPair.draftId)

    return tx.commit({tag: 'document.publish', visibility: 'async'})
  },
}
