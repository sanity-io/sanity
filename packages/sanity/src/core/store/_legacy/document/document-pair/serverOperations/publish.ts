import {type OperationImpl} from '../operations/index'
import {isLiveEditEnabled} from '../utils/isLiveEditEnabled'

type DisabledReason = 'LIVE_EDIT_ENABLED' | 'ALREADY_PUBLISHED' | 'NO_CHANGES'

export const publish: OperationImpl<[], DisabledReason> = {
  disabled: ({schema, typeName, snapshots}) => {
    if (isLiveEditEnabled(schema, typeName)) {
      return 'LIVE_EDIT_ENABLED'
    }
    if (!snapshots.draft) {
      return snapshots.published ? 'ALREADY_PUBLISHED' : 'NO_CHANGES'
    }
    return false
  },
  execute: ({client, idPair, snapshots}) => {
    // The editor must be able to see the draft they are choosing to publish.
    if (!snapshots.draft) {
      throw new Error('cannot execute "publish" when draft is missing')
    }

    return client.observable.action(
      {
        actionType: 'sanity.action.document.publish',
        draftId: idPair.draftId,
        publishedId: idPair.publishedId,
        // The editor must be able to see the latest state of both the draft document they are
        // publishing, and the published document they are choosing to replace. Optimistic
        // locking using `ifDraftRevisionId` and `ifPublishedRevisionId` ensures the client and
        // server are synchronised.
        ifDraftRevisionId: snapshots.draft._rev,
        ifPublishedRevisionId: snapshots.published?._rev,
      },
      {
        tag: 'document.publish',
      },
    )
  },
}
