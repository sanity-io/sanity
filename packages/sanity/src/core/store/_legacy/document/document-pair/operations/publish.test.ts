import type {SanityDocument} from '@sanity/types'
import {createMockSanityClient} from '../../../../../../../test/mocks/mockSanityClient'
import {isLiveEditEnabled} from '../utils/isLiveEditEnabled'
import {publish} from './publish'
import {OperationArgs} from './types'

jest.mock('../utils/isLiveEditEnabled', () => ({isLiveEditEnabled: jest.fn()}))

beforeEach(() => {
  ;(isLiveEditEnabled as jest.Mock).mockClear()
})

describe('publish', () => {
  describe('disabled', () => {
    // kind of a useless test but preserves the order at least
    it('returns with LIVE_EDIT_ENABLED if isLiveEditEnabled', () => {
      ;(isLiveEditEnabled as jest.Mock).mockImplementation(
        // eslint-disable-next-line max-nested-callbacks
        () => true,
      )

      expect(
        publish.disabled(
          // eslint-disable-next-line @typescript-eslint/no-explicit-any
          {} as any,
        ),
      ).toBe('LIVE_EDIT_ENABLED')
    })

    it('returns ALREADY_PUBLISHED if there is no draft and there is a published version', () => {
      ;(isLiveEditEnabled as jest.Mock).mockImplementation(
        // eslint-disable-next-line max-nested-callbacks
        () => false,
      )

      expect(
        publish.disabled({
          typeName: 'blah',
          snapshots: {
            draft: undefined,
            published: {} as SanityDocument,
          },
        } as unknown as OperationArgs),
      ).toBe('ALREADY_PUBLISHED')
    })

    it('returns NO_CHANGES if there is no draft and there is no published version', () => {
      ;(isLiveEditEnabled as jest.Mock).mockImplementation(
        // eslint-disable-next-line max-nested-callbacks
        () => false,
      )

      expect(
        publish.disabled({
          typeName: 'blah',
          snapshots: {
            draft: undefined,
            published: undefined,
          },
        } as unknown as OperationArgs),
      ).toBe('NO_CHANGES')
    })

    it("otherwise the operation isn't disabled", () => {
      ;(isLiveEditEnabled as jest.Mock).mockImplementation(
        // eslint-disable-next-line max-nested-callbacks
        () => false,
      )

      expect(
        publish.disabled({
          typeName: 'blah',
          snapshots: {
            draft: {} as SanityDocument,
            published: {} as SanityDocument,
          },
        } as unknown as OperationArgs),
      ).toBe(false)
    })
  })

  describe('execute', () => {
    it('removes the `_updatedAt` field', () => {
      const client = createMockSanityClient()

      publish.execute({
        client,
        idPair: {
          draftId: 'drafts.my-id',
          publishedId: 'my-id',
        },
        snapshots: {
          draft: {
            _createdAt: '2021-09-14T22:48:02.303Z',
            _rev: 'exampleRev',
            _id: 'drafts.my-id',
            _type: 'example',
            _updatedAt: '2021-09-14T22:48:02.303Z',
            newValue: 'hey',
          },
        },
      } as unknown as OperationArgs)

      expect(client.$log).toMatchSnapshot()
    })

    it('calls createOrReplace with _revision_lock_pseudo_field_ if there is an already published document', () => {
      const client = createMockSanityClient()

      publish.execute({
        client,
        idPair: {
          draftId: 'drafts.my-id',
          publishedId: 'my-id',
        },
        snapshots: {
          draft: {
            _createdAt: '2021-09-14T22:48:02.303Z',
            _rev: 'exampleRev',
            _id: 'drafts.my-id',
            _type: 'example',
            _updatedAt: '2021-09-14T22:48:02.303Z',
            newValue: 'hey',
          },
          published: {
            _createdAt: '2021-09-14T22:48:02.303Z',
            _rev: 'exampleRev',
            _id: 'drafts.my-id',
            _type: 'example',
            _updatedAt: '2021-09-14T22:48:02.303Z',
          },
        },
      } as unknown as OperationArgs)

      expect(client.$log).toMatchSnapshot()
    })

    it('takes in any and strengthens references where _strengthenOnPublish is true', () => {
      const client = createMockSanityClient()

      publish.execute({
        client,
        idPair: {
          draftId: 'drafts.my-id',
          publishedId: 'my-id',
        },
        snapshots: {
          draft: {
            _createdAt: '2021-09-14T22:48:02.303Z',
            _id: 'drafts.my-id',
            _rev: 'exampleRev',
            _type: 'my-type',
            _updatedAt: '2021-09-14T22:48:02.303Z',
            simpleRef: {
              _type: 'reference',
              _weak: true,
              _ref: 'my-ref',
              _strengthenOnPublish: true,
            },
            notToBeStrengthened: {
              _type: 'reference',
              _weak: true,
              _ref: 'my-ref',
            },
            inAn: [
              {
                _type: 'reference',
                _weak: true,
                _ref: 'my-ref-in-an-',
                _strengthenOnPublish: true,
                _key: 'my-key',
              },
              {
                _key: 'my-other-key',
                _type: 'nestedObj',
                myRef: {
                  _weak: true,
                  _ref: 'my-ref-in-an--nested',
                  _strengthenOnPublish: true,
                },
              },
              {
                _type: 'reference',
                _weak: true,
                _ref: 'my-ref-in-an--no-key',
                _strengthenOnPublish: true,
              },
            ],
          },
          published: null,
        },
      } as unknown as OperationArgs)

      expect(client.$log).toMatchSnapshot()
    })
  })
})
