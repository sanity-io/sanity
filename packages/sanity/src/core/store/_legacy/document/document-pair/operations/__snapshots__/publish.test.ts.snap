// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`publish > execute > calls createOrReplace with _revision_lock_pseudo_field_ if there is an already published document 1`] = `
{
  "listen": [],
  "observable": {
    "action": [],
    "fetch": [],
    "getDocuments": [],
    "listen": [],
    "request": [],
  },
  "request": [],
  "transaction": [
    {
      "commit": [
        [
          {
            "tag": "document.publish",
            "visibility": "async",
          },
        ],
      ],
      "create": [],
      "createIfNotExists": [],
      "createOrReplace": [
        [
          {
            "_createdAt": "2021-09-14T22:48:02.303Z",
            "_id": "my-id",
            "_rev": "exampleRev",
            "_type": "example",
            "newValue": "hey",
          },
        ],
      ],
      "delete": [
        [
          "drafts.my-id",
        ],
      ],
      "id": 1,
      "patch": [
        [
          "my-id",
          {
            "ifRevisionID": "exampleRev",
            "unset": [
              "_revision_lock_pseudo_field_",
            ],
          },
        ],
      ],
    },
  ],
}
`;

exports[`publish > execute > removes the \`_updatedAt\` field 1`] = `
{
  "listen": [],
  "observable": {
    "action": [],
    "fetch": [],
    "getDocuments": [],
    "listen": [],
    "request": [],
  },
  "request": [],
  "transaction": [
    {
      "commit": [
        [
          {
            "tag": "document.publish",
            "visibility": "async",
          },
        ],
      ],
      "create": [
        [
          {
            "_createdAt": "2021-09-14T22:48:02.303Z",
            "_id": "my-id",
            "_rev": "exampleRev",
            "_type": "example",
            "newValue": "hey",
          },
        ],
      ],
      "createIfNotExists": [],
      "createOrReplace": [],
      "delete": [
        [
          "drafts.my-id",
        ],
      ],
      "id": 1,
      "patch": [],
    },
  ],
}
`;

exports[`publish > execute > takes in any and strengthens references where _strengthenOnPublish is true 1`] = `
{
  "listen": [],
  "observable": {
    "action": [],
    "fetch": [],
    "getDocuments": [],
    "listen": [],
    "request": [],
  },
  "request": [],
  "transaction": [
    {
      "commit": [
        [
          {
            "tag": "document.publish",
            "visibility": "async",
          },
        ],
      ],
      "create": [
        [
          {
            "_createdAt": "2021-09-14T22:48:02.303Z",
            "_id": "my-id",
            "_rev": "exampleRev",
            "_type": "my-type",
            "inAn": [
              {
                "_key": "my-key",
                "_ref": "my-ref-in-an-",
                "_type": "reference",
              },
              {
                "_key": "my-other-key",
                "_type": "nestedObj",
                "myRef": {
                  "_ref": "my-ref-in-an--nested",
                },
              },
              {
                "_ref": "my-ref-in-an--no-key",
                "_type": "reference",
              },
            ],
            "notToBeStrengthened": {
              "_ref": "my-ref",
              "_type": "reference",
              "_weak": true,
            },
            "simpleRef": {
              "_ref": "my-ref",
              "_type": "reference",
            },
          },
        ],
      ],
      "createIfNotExists": [],
      "createOrReplace": [],
      "delete": [
        [
          "drafts.my-id",
        ],
      ],
      "id": 1,
      "patch": [],
    },
  ],
}
`;
